Use GFX_* functions for splash screen and fix SDL texture related issues

dreamer opened the pull request:
(Only a draft for now, as some more changes need to land)

- Use GFX_StartUpdate/EndUpdate to draw splash screen with user-selected `output` pref; this makes it possible to draw splash in fullscreen, so there's no awkward transition from fullscreen to splash window to fullscreen when starting via Steam Big Picture. This is also pre-requisite towards #139 and fixes #178 as a side-effect.
- Use updated splash screen of size 640x480 (better-compressed png file as source, prevents window resize with dosbox-staging default startup values).
- Replace SDL_Lock/Unlock texture with SDL_UpdateTexture: fixes #149 (at least on Windows, hopefully on macOS as well), improves stability on AmigaOS: #177.
- Code cleanups around this code, fixing #180 to improve stability on Linux+Wayland and AmigaOS.

krcroft commented:
This is looking great @dreamer ; thrashing between video modes on startup (on PPC with a slow CPU, the window would flicker in and out three to four times before the game gets underway), but now startup flicker is gone and the transition between the splash to the game feels solid.

I ran through a lot of scenarios; only able to identify one potential corner case (but might be my system).  Try this:

```
[sdl]
fullscreen       = true
vsync            = true
fullresolution   = desktop
windowresolution = 1280x960
output           = texturepp
renderer         = opengl
```

On my Linux system, the splash does not show, the window appears to hang momentarily, and then the game gets underway.  Also occurs w/ `output = texture` and  texturenp`.

dreamer commented:
> ```
> output           = texturepp
> renderer         = opengl
> ```
`texture_renderer` ;) I presume this will be a common mistake, I should probably add handling and warning log msg when `renderer` is used. Aside of that, I see `vsync` is turned to true - I haven't tested with vsync enabled due to our other vsync-related problems - I'll check it out.

Personally I still see some flickering (but not always - I see it on my Desktop, with AMD GPU, and I don't see it on my laptop with Intel GPU); it's **much** more visible on Windows 10 than on Linux, and I haven't been able to reach the root cause of that flicker yet (happens only with OpenGL; Windows+texture+directx does not result in flicker…).

dreamer commented:
Uff, I think I'm finished :)

I know the number of changed lines indicated in the review is massive, but it's largely due to adjusted splash screen image, in reality changes span ~650 lines (which is a lot, anyway):
```
$ git diff --stat master..HEAD
 README                                |   12 -
 include/control.h                     |   39 +-
 include/render.h                      |    2 +-
 include/support.h                     |    5 +
 include/video.h                       |    9 +-
 src/dosbox_staging_splash_640_400.png |  Bin 28761 -> 0 bytes
 src/dosbox_staging_splash_640_480.png |  Bin 0 -> 19997 bytes
 src/gui/dosbox_splash.h               | 1028 -----------------
 src/gui/dosbox_staging_splash.c       | 3011 +++++++++++++++++++++++-------------------------
 src/gui/sdlmain.cpp                   |  580 +++++-----
 src/libs/ppscale/ppscale.h            |    8 +
 11 files changed, 1785 insertions(+), 2909 deletions(-)
```

I'll start with the "bad" news:

- @krcroft I was not able to reproduce the problem you mentioned earlier :( the last idea I have about this is - maybe try removing ifdef in sdlmain.cpp:556, so the same hack I developed around crash in Windows OpenGL drivers will apply to your build. It's a long shot, though.
- as I was pulling the threads, I discovered more additional small bugs in weird corner cases - but I didn't want to keep piling on fixes; nothing really critical though
- this PR increases number of effc++ warnings :( - all due to me changing `render_driver` field in SDL_block to `std::string` - compiler warns us, that all those fields should be initialized now. Whenever touching a field in SDL_block I added an initializer, but that nested structure is huge, so decided not to do it for all fields at once. I hope we'll just solve it over time.

Now, the good news:

- Fedora updated SDL to 2.0.12 - and in result, the last weird flicker I was experiencing is now gone :) @krcroft I hope update will solve the issue you mentioned as well; now we need to wait for update to land in vcpkg and brew
- I figured out why `output=surface` resulted in broken splash screen (but only when in fullscreen), and it's now fixed in c8cbac4 
- I remember the discussion from Vogons about user who couldn't run dosbox in low resolution because the splash screen was bigger than his old CRT could handle - I managed to reproduce similar problem, and with splash refactored to separate function, the fix was trivial: a5ce551
- Working with cross-platform app and testers on various OSes is great :) @kas1e suggested a fix for AmigaOS, which solved a problem on Windows 10, which made me notice an assert in Windows debug builds, solving of which triggered GCC sanitizer on Linux to indicate an indirect memory leak which was affecting all the platforms. This is great :D

dreamer commented:
Forgot to mention: I recommend doing review commit-by-commit; if we can make it smaller by merging some changes earlier that would be nice, but at the moment I can think only of 1-2 commits in here, that could be easily extracted.

krcroft commented:
Just giving it some permutation testing right now; and overall the transition between modes is solid - this is a huge improvement to the SDL2 portion of dosbox-staging! 

I /think/ I've found the combination to reproduce the no-banner, no-help screen issue.  Drop this `dosbox.conf` into your Rayman directory.  It may take two or three launches to reproduce. 

 [dosbox.conf.txt](https://github.com/dreamer/dosbox-staging/files/4364479/dosbox.conf.txt)

I believe this is the minimal set. 
```
[sdl]
fullscreen       = true
vsync            = true
output           = texturepp
texture_renderer = opengl

[render]
aspect   = false
```
Regarding `vsync` - is this something that changes or improves games? (just an honest question; not casting doubt against it)  I know from our testing, it not really useful on Linux -  but might be on other OSes?  If it's still in limbo, we could comment out that block from the `conf` and nix it until someone figures it out.  One less semi-harmful option in the `conf` is an improvement. (could also be a separate on-off PR too)

krcroft commented:
Regarding `vsync` - is this something that changes or improves games? (just an honest question; not casting doubt against it)  I know from our testing, it not really useful on Linux -  but might be on other OSes?  If it's still in limbo, we could comment out that block from the `conf` and nix it until someone figures it out.  One less semi-harmful option in the `conf` is an improvement. (could also be a separate PR too)

krcroft commented:
Interesting; when I leave this section enabled like you suggested:
```
#if defined (WIN32)
		// This is a hack for Windows 10 to prevent a crash in AMD OpenGL
		// driver when window is being re-created by SDL2 internally to
		// support OpenGL.
		else if (screenType == SCREEN_TEXTURE && starts_with("opengl", sdl.render_driver))
			flags |= SDL_WINDOW_OPENGL;
#endif
```

I can generate the "splash-stall" in Windowed mode too using this conf file:

[dosbox-windowed-start-stall.conf.txt](https://github.com/dreamer/dosbox-staging/files/4364592/dosbox-windowed-start-stall.conf.txt)

The content area becomes "transparent" and the window freezes for the duration of the splash screen, then proceeds straight to the game startup:

![2020-03-21_20-07](https://user-images.githubusercontent.com/1557255/77241467-c19d9400-6baf-11ea-9c98-1aee7d274882.png)

krcroft commented:
Likely unrelated to this PR, but with:

```
[sdl]
fullscreen       = false
windowresolution = desktop
output           = opengl
texture_renderer = opengl
```

The window isn't hardware-scaled to the desktop size:

[dosbox-windows-desktop-res-too-small.conf.txt](https://github.com/dreamer/dosbox-staging/files/4364601/dosbox-windows-desktop-res-too-small.conf.txt)

![2020-03-21_20-21](https://user-images.githubusercontent.com/1557255/77241597-a16ed480-6bb1-11ea-98db-b81c447a1dd5.png)

However, if I hardcore a Window resolution then HW-scaling and PP will scale up:
```
[sdl]
fullscreen       = false
windowresolution = 1280x960
output           = opengl
texture_renderer = opengl
```

dreamer commented:
> Regarding `vsync` - is this something that changes or improves games? (just an honest question; not casting doubt against it) I know from our testing, it not really useful on Linux - but might be on other OSes? If it's still in limbo, we could comment out that block from the `conf` and nix it until someone figures it out. One less semi-harmful option in the `conf` is an improvement. (could also be a separate PR too)

This is *the* first option people tend to look for when they notice tearing in a game.

I never managed to make it work in DOSBox, ever - I experience tearing regularly, in most games emulated via any dosbox version I ever tried. I don't know why, but all games I try to run in dosbox (any fork) are capped at max 70 FPS (DOS in-game FPS counters might say more, but FPS counter hooking into OpenGL say otherwise), while my displays are 60Hz. Maybe 70Hz was the popular refresh rates for CRTs? I don't know… but, this max FPS lock guarantees tearing for me.

I am ok with removing `vsync` option from settings until we (or someone else) will figure out why it does not work.


> `windowresolution = desktop`

Was this option ever supported? Right now it falls-back to the default value / `original`. Since recently, dosbox-SVN supports setting window size as percentage (it's undocumented feature) - so `desktop` could be equivalent of `windowresolution = 100%`). When merging with SVN code I removed handling of percentage, because it was totally conflicting with SDL2 code (but I can re-implement it back, it's a small feature).

So, should I implement window size as % and make `desktop` as an alias for `windowresolution=100%`?

> (reproduction steps)

No luck with reproducing the issue so far… :( And the 1-frame flicker I mentioned earlier as fixed with SDL 2.0.12 came back (no idea what triggers it - maybe it's differnce between X11 and XWayland).

But I need to mention a different issue, that I noticed earlier with different games as well (perhaps we'll need to create a new discussion-issue for it) - when testing Rayman, `cycles = fixed 30000` results in waaay too fast game on my computer - does it work ok on yours? I compared to gameplay videos on YouTube, and through trial and error arrived at `cycles = fixed 15000`. According to Qbix, setting fixed cycles should be "portable", but it never was in my experience (in any dosbox fork).

krcroft commented:
> should I implement window size as % and make desktop as an alias for windowresolution=100%?

That would work; but sounds like the percent doesn't play nice with SDL2. 
Maybe easier to have `desktop` query the desktop size. Something like:

(if `conf` fullscreen is false) ...

```
SDL_DisplayMode dm;
if (SDL_GetDesktopDisplayMode(0, &dm) != 0)
{
     LOG_MSG("SDL_GetDesktopDisplayMode failed: %s", SDL_GetError());
     return 1;
}
// Now we can use:
dm.w;
dm.h;
```

krcroft commented:
> No luck with reproducing the issue so far… :( 
Darn! Let's shelve it and consider it something specific to my environment.  Plus, it will disappear if we drop vsync too (which I agree with).

krcroft commented:
> Maybe 70Hz was the popular refresh rates for CRTs

Yup; it was dictated by video mode. A subset of games use 60Hz - I will try to find that kist. These should be mostly tear free.. but I'm not sure it guarantees it.. we need double buffering so the next frame is fully rendered and ready to flip to, prior to the next screen update.

Some games only render at a subset of 70Hz (one of the ID games uses 35 Hz).

krcroft commented:
> when testing Rayman, cycles = fixed 30000 results in waaay too fast gameplay

Oh; definitely not.  Rayman is very "well behaved" like most protected mode games.  I can drive the cycles upward of 1M without any change in play-speed.  

At very low cycles (say, 3000) protected mode games are choppy - and then as I increase cycles they approach their ideal speed at which point giving them more simply will waste cycles without improving smoothness or frame-rates.  The game itself still play at correct speed.

_(Sure, there are plenty of real-mode games from the 80s that are very affected by cycles; but that's true on hardware + real DOS too.  And some protected mode games that don't have internal timers, like Hi-Octane; so they speed away on fast hardware. )_

But yeah, I agree with Qbix too; cycles are full portable for me - across hardware types and even core types (dynamic vs normal). 

I wonder if it's Wayland? Or a video driver setting?

dreamer commented:
Ugh, when I look at the number of commits here… @krcroft I think I can split away few cleanup commits to a separate review, to make this one smaller. What do you think?

Also, I would prefer to work on `windowsize=X%` and `windowsize=desktop` in a separate PR after this one.

krcroft commented:
If the commits are already package up by theme or individual features, then I'm OK with one big review too; because either way we can bisect down if corner cases pop up.  Sounds good to move resolution=desktop to the next PR.  Will start reviewing in detail tomorrow!

dreamer commented:
I discovered what was the cause of "flicker" I mentioned earlier. Turns out it is correlated to Gnome 3 disabling compositing for a fullscreen application, I will need to discuss it with Mutter and Mesa developers to understand how to prevent it (or where to fix it). This flicker is not really breaking anything, just a tiny graphical artefact, so out of scope for now.

There are two workarounds to prevent flicker it from happening:
- Using [this](https://extensions.gnome.org/extension/1873/disable-unredirect-fullscreen-windows/) Gnome extension
- Switching to Wayland and running dosbox-staging via XWayland

@krcroft I will do some testing using KDE - maybe the problem with transparent splash screen depends on compositor as well?

krcroft commented:
> I will do some testing using KDE - maybe the problem with transparent splash screen depends on compositor as well?

Definitely could be; I'm using MATE and looks like the compositor was enabled (not sure if it's disabled when going fullscreen or not); but I can disable the compositor entirely, so will check that too.

krcroft commented on include/setup.h:
```diff
@@ -111,26 +111,41 @@ class Value {
-	Property(std::string const& _propname, Changeable::Value when):propname(_propname),change(when) { }
+	Property(const std::string &name, Changeable::Value when)
+		: propname(name),
+		  value(),
```
Is an empty 'name' string ok? could add a check in the constructor (I'm not sure if the origin code checked, but just thinking defensively).

dreamer replied:
Yeah, we should prevent empty name - just in case. Perhaps I can code it in a way to test for this case at compile time - I think properties name are always hardcoded during definition.

dreamer replied:
Ah, I went into the rabbit hole of trying to handle std::string in a `constexpr`… (this is impossible due to C++ design of std::string class). Decided to simply place an `assertm` in the constructor - if anyone will program empty property name, it will be obvious on the first test.

krcroft commented on include/support.h:
```diff
@@ -122,4 +122,9 @@ void upcase(std::string &str);
+bool starts_with(const char (& pfx)[N], const std::string &str) noexcept {
+	return (strncmp(pfx, str.c_str(), N) == 0);
+}
+
```
Another nice addition!

dreamer replied:
I just miss Python's sensibly named functions for operating on strings :) I considered adding `bool starts_with(const std::string &, const std::string&)` as well, but that function would be left unused, so this variant is enough for now.

krcroft commented on src/gui/dosbox_staging_splash.c:
```diff
@@ -22,13 +22,13 @@
-  640, 400, 3,
+  unsigned char	 rle_pixel_data[39136 + 1];
+} splash_image = {
+  640, 480, 3,
```
Would it make sense to rendering the SVG on the fly using the single-header NanoSVG:
- https://github.com/memononen/nanosvg
- https://www.gamedev.net/forums/topic/652219-resolution-independent-graphics-with-svg-in-sdl/5125198/

Entirely cosmetic; but when you've got a logo this good, it might as well be rendered to the max :-)

Also, thinking about binary size, NanoSVG might compile down to something roughly the same size as carrying around these embedded graphics anyway.. and the benefit will be beautiful logos on latest 4k panels downt to old 1024x768 displays.

dreamer replied:
Great suggestion! It crossed my mind long time ago, but I looked at SDL_image back then and decided to leave it be. Haven't heard about nanosvg before - definitely worth considering.

Integrating SVG support can be also seen as a prerequisite for developing better GUI - I was thinking about replacing mapper GUI with proper, nice settings screen. Most DOSBox options can be switched during runtime (DOSBox-X uses this a lot).

krcroft commented on src/gui/sdlmain.cpp:
```diff
@@ -195,6 +194,19 @@ extern char** environ;
+#if SDL_BYTEORDER == SDL_BIG_ENDIAN
+constexpr uint32_t RMASK = 0xff000000;
+constexpr uint32_t GMASK = 0x00ff0000;
+constexpr uint32_t BMASK = 0x0000ff00;
```
This should cover the big-endian PPC non-macOS systems like @kas1e and I have. 
For macOS systems running big-endian processors, like the G5 - do they use little-endian when it comes to SDL?

dreamer replied:
I think it will be a bit clearer when looking at context of this change: 0b87c0bc6a0a88a33ea6dbefa2b073faeddeda48

In old code, these masks were used in 2 contexts:
1. For rendering icon used as a fallback when icon is not set other way. This code was ifdefed out for macOS, with an explanation that AppBundle includes nicer icon, so this one is not needed… This explanation is only partly true, but *shrug* - I decided to preserve the behaviour.  *Side note: this code was also broken in SVN for who-knows-how-long, and the fallback icon was not showing up on Linux - but now it does :)*
2. For rendering warning messages in a window (on Windows only) - I have never ever seen this code in action on any Windows build, but it's there. Not sure why PPC ifdefs were appearing in Windows-specific code (maybe it was used on OS/2 as well in the past?) No idea.

So, after moving these masks out of function scopes, our CI hit me with a new warning: unused variable, only on macOS builds.

krcroft commented on src/gui/sdlmain.cpp:
```diff
@@ -243,21 +256,20 @@ struct SDL_Block {
-		bool lazy_fullscreen;
-		bool lazy_fullscreen_req;
-		bool vsync;
+		bool vsync = false;
```
We might not even need vsync, because we're implicitly using double-buffering by using [SDL_RenderPresent](https://wiki.libsdl.org/SDL_RenderPresent), which describes writing to a back-buffer that's then flipped to and rendered.

dreamer replied:
I still think we need vsync - e.g. for supporting adaptive sync in the future (it's the same setting basically, just a different value).

krcroft replied:
Good stuff.  We can atleast shield the user from dealing with vsync; and manager this required sdl option internally until it proves useful to expose to the user.

krcroft commented on src/gui/sdlmain.cpp:
```diff
@@ -243,21 +256,20 @@ struct SDL_Block {
 		SDL_GLContext context;
-		Bitu pitch;
+		int pitch = 0;
 		void * framebuf; // TODO Use unique_ptr to prevent leak
```
`framebuf = nullptr`?

dreamer replied:
Done :) Thanks for pointing this out. I also removed TODO comment, as this buffer should not be managed via unique_ptr after all.

krcroft commented on src/gui/sdlmain.cpp:
```diff
@@ -519,84 +524,72 @@ static int int_log2 (int val) {
+		// support OpenGL.
+		else if (screenType == SCREEN_TEXTURE && starts_with("opengl", sdl.render_driver))
+			flags |= SDL_WINDOW_OPENGL;
+#endif
```
Is it possible to move this into the config-file code? 

For example, if this hack limits Windows systems to using a particular subset of the `output` and `texture_renderer` values, maybe we can move that limitation to the config-file code - which would make it impossible for Windows users to set bad combinations (and it would streamline this part).

dreamer replied:
Not sure how it would work… the code depends on a runtime state and is needed only on windows. Right now Windows users cannot set bad combination - because of `starts_with("opengl", …)` the code handles `opengl`, `opengles`, `opengles2`, etc.

Hopefully I'll manage to figure out and prevent this hack in the future (somehow), so I would prefer not to propagate it to other source files.

krcroft replied:
Thanks for explaining the dependencies; makes sense!

krcroft commented on src/gui/sdlmain.cpp:
```diff
@@ -519,84 +524,72 @@ static int int_log2 (int val) {
+		// Using undefined position will take care of placing and restoring the
+		// window by WM.
+		const int sdl_pos = SDL_WINDOWPOS_UNDEFINED_DISPLAY(sdl.displayNumber);
+		sdl.window = SDL_CreateWindow("", sdl_pos, sdl_pos, width, height, flags);
```
Really nice improvement; let the WM handle placement (allow position to saved, etc..)

dreamer replied:
Yup works fine :) There's one corner case when it does not work, but after working on it for some time - I strongly suspect it's a tiny bug in SDL2; I need to consult SDL devs about that.

The corner case is: when starting in fullscreen mode, and then leaving with alt-enter, window is always placed at 0,0 (I preserved a hack for windows to work around this though); this should not be the case, but I think that SDL2 calculates new window position using window size equal to fullscreen resolution.

In previous PR I developed a hack to avoid it on Linux as well, but this PR removes that - it was causing problems in multi-desktop environment.

krcroft commented on src/gui/sdlmain.cpp:
```diff
@@ -1388,77 +1381,66 @@ void GFX_SwitchFullScreen(void) {
-		pitch = texPitch;
-		sdl.updating=true;
+		pixels = static_cast<uint8_t *>(sdl.texture.input_surface->pixels);
+		pitch = sdl.texture.input_surface->pitch;
```
`assert(sdl.texture.input_surface)` before 1410 to prevent null dereference? (not sure if it needed though).

dreamer replied:
Yes! :) Using these functions in a new context without assertions and without documentation was very inconvenient. I added all asserts suggested below as well.

krcroft commented on src/gui/sdlmain.cpp:
```diff
@@ -1388,77 +1381,66 @@ void GFX_SwitchFullScreen(void) {
+	case SCREEN_SURFACE:
+		pixels = static_cast<uint8_t *>(sdl.surface->pixels);
+		pixels += sdl.clip.y * sdl.surface->pitch;
+		pixels += sdl.clip.x * sdl.surface->format->BytesPerPixel;
```
`assert(sdl.surface)` before 1432?

krcroft commented on src/gui/sdlmain.cpp:
```diff
@@ -1467,28 +1449,11 @@ void GFX_EndUpdate( const Bit16u *changedLines ) {
-		break;
 	case SCREEN_TEXTURE:
-		SDL_UnlockTexture(sdl.texture.texture);
+		SDL_UpdateTexture(sdl.texture.texture,
```
`assert(sdl.texture.input_surface)` before 1453? (I suppose these are only needed if the static analyzer flags these.. I'm just been unnecessarily defensive)

krcroft commented on src/gui/sdlmain.cpp:
```diff
@@ -2335,22 +2349,25 @@ void Config_Add_SDL() {
-	Pbool->Set_help("Start dosbox directly in fullscreen. (Press ALT-Enter to go back)");
+	constexpr auto always = Property::Changeable::Always;
+	constexpr auto deprecated = Property::Changeable::Deprecated;
+
```
Nice short-hand simplification!

dreamer replied:
Hey, you added the `always` shortcut in here originally ;) It is convenient.

dreamer commented:
Applied review requests (via fixups to the commits where surrounding code was implemented) and rebased - thanks for the review :) if everything is ok, I'll merge with next approval.

I made a round of quick tests on Cinnamon, MATE and KDE Plasma - without noticing any problems. I did found a new small alt-tab related bug, but it's not a regression (reproducible only in Gnome, also in other dosbox forks).

krcroft commented:
Every is good from my testing too - behaves equal to or better (without startup flicker) than before.  Ship it :-)

dreamer commented:
Done. Thanks! :)

Up next: resizable window and support for `windowsize=X%`. These are the last two bigger features I want to land for 0.75.0.
